runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/WeightedBoundedQueue.java

```diff
@@ -60,14 +64,29 @@ public void put(V value) {
    * Retrieves and removes the head of this queue, waiting up to the specified wait time if
    * necessary for an element to become available.
    *
-   * @param timeout how long to wait before giving up, in units of {@code unit}
-   * @param unit a {@code TimeUnit} determining how to interpret the {@code timeout} parameter
+   * @param timeout how long to wait before giving up
    * @return the head of this queue, or {@code null} if the specified waiting time elapses before an
    *     element is available
    * @throws InterruptedException if interrupted while waiting
    */
-  public @Nullable V poll(long timeout, TimeUnit unit) throws InterruptedException {
-    @Nullable V result = queue.poll(timeout, unit);
+  public @Nullable V poll(Duration timeout) throws InterruptedException {
+    @Nullable V result;
+    Instant deadline = Instant.now().plus(timeout);
+    int spin = 0;
+    while (true) {
+      if (++spin > 1000) {
```

Comment:
how much cpu does this burn if the queue is empty?

we could also look into other specialized queues like:
https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/MpscBlockingConsumerArrayQueue.java
(random googling, would need some validation).